Add extend packer build template engine docs (#8615)
hashicorp/packer: website/source/docs/extending/custom-builders.html.md

and will likely change in a future version. They aren't fully "baked" yet, so
they aren't documented here other than to tell you how to hook in provisioners.

## Template Engine

> 

Reviewer: Here I'd may be add a subtitle:

```suggestion
### Build variables
```

Reviewer 2: :+1:

Packer makes it possible to provide custom template engine variables to be shared along 
provisioners using the `build` function. 
 
Part of the builder interface changes made in 1.5.0 was to make builder Prepare() methods 
return a list of custom variables which we call `generated data`. 
We use that list of variables to generate a custom placeholder map per builder that 
combines custom variables with the placeholder of existing variables created by Packer.   

> An example of returning generated date placeholder inside Prepare() method:

Reviewer: ```suggestion
An example of returning generated data placeholder inside Prepare() method:
```

> An example of returning generated date placeholder inside Prepare() method:

Reviewer 2: ```suggestion
An example of returning generated data placeholder inside Prepare() method:
```

```
func (b *Builder) Prepare(raws ...interface{}) ([]string, []string, error) {
    // custom prepare code
    
    generatedData := []string{"SourceImageName"}
    return generatedData, warns, nil
}
```
   
Once the placeholder is set, it's necessary to pass the variables real values when calling 
the provisioner. This can be done as the example below:   

```
func (b *Builder) Run(ctx context.Context, ui packer.Ui, hook packer.Hook) (packer.Artifact, error) {
    // custom run code
    
    // Create map of custom variable

>     generatedData := map[string]interface{}{"SourceImageName": buildInfoTemplate.SourceImageName}

Reviewer: ```suggestion
    generatedData := map[string]interface{}{"SourceImageName": "the source image name value"}
```

    // Pass map to provisioner
    hook.Run(context.Context, packer.HookProvision, ui, comm, generatedData)
    

>     // custom run code

Reviewer: ```suggestion
    // ...
```

and will likely change in a future version. They aren't fully "baked" yet, so
they aren't documented here other than to tell you how to hook in provisioners.

## Template Engine

### Build variables

> Packer makes it possible to provide custom template engine variables to be shared along 

Reviewer 3: ```suggestion
Packer makes it possible to provide custom template engine variables to be shared with 
```

provisioners using the `build` function. 
 
Part of the builder interface changes made in 1.5.0 was to make builder Prepare() methods 
return a list of custom variables which we call `generated data`. 
We use that list of variables to generate a custom placeholder map per builder that 

> combines custom variables with the placeholder of existing variables created by Packer.   

Reviewer 2: ```suggestion
combines custom variables with the placeholder map of default build variables created by Packer.   
```

Here's an example snippet telling packer what will be made available by the builder:

```
func (b *Builder) Prepare(raws ...interface{}) ([]string, []string, error) {
    // ...
    
    generatedData := []string{"SourceImageName"}
    return generatedData, warns, nil
}
```
Returning the custom variable name(s) within the `generated_data` placeholder is necessary 
for the template containing the build variable(s) to validate.  
    

> Once the placeholder is set, it's necessary to pass the variables real values when calling 

Reviewer 3: ```suggestion
Once the placeholder is set, it's necessary to pass the variables' real values when calling 
```
